Expand structure and signature items before applying attribute rules

NathanReb opened the pull request:
This PR changes the order in which context-free rules are applied.

It is a follow up to #276 and "fixes" the behaviour observed there. With this PR a node with attributes attached will be first expanded and then the inline attribute rules applied instead of the other way around.

Imagine the following:
```ocaml
type t = [%some_ext ...]
[@@deriving some_deriver]
```
Before this PR, the `some_deriver` transformation would be applied first, now it is applied to the result of the expansion of the type declaration, meaning after the `some_ext` transformation was applied.

@kit-ty-kate is gracefully testing this on the opam universe to see if it has any serious impact. I don't expect anyone to rely on this behaviour but better be safe than sorry.

Edit: You might want to review this while hiding whitespace diffs as a whole block in `Context_free` has gained one level of identation without any further change.

kit-ty-kate reviewed (approved):
The check in opam-repository I did, didn't reveal any packages broken by this

NathanReb commented:
I updated the internal documentation and the tests description to reflect this change. I think this should be ready for review now.

@jeremiedimino @ceastlund, I don't really expect anything to break here but it might be good to test this change inside JS' monorepo. What do you think?

NathanReb commented:
Rebased on top of (formatted!) main and fixed the bug found by @pitag-ha where I forgot to remove a `self#structure_item` call in `Context_free.map_top_down#structure` resulting in some structure item being traversed twice.

pitag-ha reviewed (approved):
Looks good also to me. Thanks!
